### PR TITLE
LEARNER-6679 Updated ProgramApplicator to CustomApplicator.

### PR DESCRIPTION
**Pull request by Ayub-Khan**
-Override Applicator get_site_offers to only get offers
valid for baket without bundle.
-Override BasketMiddleware apply_offers_to_basket to
use CustomApplicator.
-Used CustomApplicator in basketmiddleware and utils
-Updated constants
-Updated tests

### CODE REVIEW

File: ecommerce/extensions/offer/constants.py [hunk 1]
```diff
@@ -1,2 +1,2 @@
-PROGRAM_APPLICATOR_LOG_FLAG = 'enable_program_applicator_logging'
-PROGRAM_APPLICATOR_USE_FLAG = 'enable_program_applicator_use'
+CUSTOM_APPLICATOR_LOG_FLAG = 'enable_program_applicator_logging'
+CUSTOM_APPLICATOR_USE_FLAG = 'enable_program_applicator_use'
```
Comment (Ayub-Khan) [inline ecommerce/extensions/offer/constants.py]: Only updating Variables as we already have flags.

File: ecommerce/extensions/basket/middleware.py [hunk 2]
```diff
@@ -62,3 +69,11 @@ def get_basket(self, request):
         request._basket_cache = basket
 
         return basket
+
+    @newrelic.agent.function_trace()
+    def apply_offers_to_basket(self, request, basket):
```
Comment (Ayub-Khan) [inline ecommerce/extensions/basket/middleware.py:74]: This is the place from where applicator is actually called for baskets without vouchers(bundle baskets) so we need to override this. 
https://github.com/django-oscar/django-oscar/blob/1.5.1/src/oscar/apps/basket/middleware.py#L202

File: ecommerce/extensions/offer/applicator.py [hunk 3]
```diff
@@ -67,6 +72,14 @@ def get_offers(self, basket, user=None, request=None):
             )
         )
 
+    def get_site_offers(self):
```
Comment (Ayub-Khan) [inline ecommerce/extensions/offer/applicator.py:75]: Overrides Applicator get_site_offers. We do not want to check bundle offers against voucher baskets.
https://github.com/django-oscar/django-oscar/blob/1.5.1/src/oscar/apps/offer/applicator.py#L63

**Comment by Ayub-Khan** [comment]
Previously if there are 4 products in program and student have bought 3 he would have got the bundle discount on the last basket in case of direct purchases(no bundle id in basket). this pr will disable that. After that student will only get discount if purchases the last product or remaining products using the bundle purchase option. If he directly purchases the product wont get any discount. 
But this will save us from 150+ program api calls to course discovery and 150-300+ enrollment and entitlement api calls to platform for every basket/add call.

File: ecommerce/extensions/offer/constants.py (hunk 1 above)
Comment (BenjiLee) [reply ecommerce/extensions/offer/constants.py]: We should also update the flags. The flags can always be renamed in the django admin.

**Comment by robrap** [comment]
@BenjiLee @Ayub-Khan @brittneyexline @georgebabey: Can you document the plan for our intention around the applicator(s)?  Given that we call the applicator in multiple places, do we need just one that knows how to respond accordingly?  Was a special one being added for Enterprise, and will that break-down when the/an applicator is used in later parts of the payment process?  If this has already been thought through and it all just works, can you help clarify to the team how it works?  If not, can you make it work and then clarify? ;)

**Comment by BenjiLee** [comment]
This PR looks fine to me. I'm not sure what the enterprise workflow looks like and whether or not it includes a bundle id in the request or if program offer can be applied WITH an enterprise offer. @brittneyexline and @georgebabey may have a better idea of if the new workflow here will affect enterprise.

**Review by BenjiLee — APPROVED** [review]

**Comment by brittneyexline** [comment]
@BenjiLee @robrap - we don't yet have plans to change the Applicator logic, so these changes should not impact our plans. If or when we do need to change the Applicator, we will update you on the plans.

**Comment by robrap** [comment]
Thanks  @brittneyexline.

**Pull request merged**